### PR TITLE
Bump machines-endpoints 0.9.5

### PR DESCRIPTION
**Pull request by takara9**

### CODE REVIEW

File: machines-endpoints/pkg/machines-endpoints/main.go [hunk 1]
```diff
@@ -418,4 +422,74 @@ func main() {
...
 		}
 	}
+
+	if *flgBMCLogCollectorConfigMap {
+		// create bmc-log-collector configmap of all servers that include boot servers
+		addersses, err := updateServerAddressList(machines)
+		if err != nil {
+			log.ErrorExit(err)
+		}
+		err = client.writeConfigMap(ctx, addersses)
```
Comment (masa213f) [inline machines-endpoints/pkg/machines-endpoints/main.go:432]: The function name `writeConfigMap()` is too abstract.
Please give it a name that clarifies the target resource.
For example, in the case of bmc-reverse-proxy, it is `client.updateBMCProxyConfigMap()`.

Also, all the functions that update the resources in this program are `client.updateHOGE()`.
But only this one is `client.writeHOGE()`. Please make the function names consistent.

File: machines-endpoints/pkg/machines-endpoints/main.go [hunk 2]
```diff
@@ -418,4 +422,74 @@ func main() {
...
+		if err != nil {
+			log.ErrorExit(err)
+		}
+	}
+}
+
+func updateServerAddressList(machines []Machine) (map[string]string, error) {
```
Comment (masa213f) [inline machines-endpoints/pkg/machines-endpoints/main.go:439]: I don't know why `updateServerAddressList()` and `client.writeConfigMap()` are defined after `main()`.

The existing functions are defined in the order of the client's receivers, other functions, and `main()`.
Please check the location where defining them.

File: machines-endpoints/pkg/machines-endpoints/main.go [hunk 3]
```diff
@@ -418,4 +422,74 @@ func main() {
...
+	type MachineConfigMap struct {
```
Comment (masa213f) [inline machines-endpoints/pkg/machines-endpoints/main.go:440]: Is this struct name `MachineConfigMap` appropriate? Is this a ConfigMap?

File: machines-endpoints/pkg/machines-endpoints/main.go [hunk 4]
```diff
@@ -418,4 +422,74 @@ func main() {
...
+	}
+
+	byteJSON, err := json.Marshal(ml)
+	if err != nil {
+		return nil, err
+	}
+	addresses := make(map[string]string)
+	addresses["serverlist.json"] = string(byteJSON)
+	return addresses, nil
+}
```
Comment (masa213f) [inline machines-endpoints/pkg/machines-endpoints/main.go:469]: IMO, it is better to return the contents of `server-list JSON`.

It is not good to depend on the structure of the ConfigMap.
Wrapping with a map for ConfigMap should be done in `writeConfigMap()`.

```suggestion
	byteJSON, err := json.Marshal(ml)
	if err != nil {
		return nil, err
	}
	string(byteJSON), nil
}
```

**Review by masa213f — CHANGES_REQUESTED** [review]
Could you remove the first step `1. Update version variables in Makefile.` in `maintenance.md`?
Now, version numbers are not written in `Makefile`.
https://github.com/cybozu/neco-containers/blob/main/maintenance.md#machines-endpoints

File: machines-endpoints/pkg/machines-endpoints/main.go [hunk 5]
```diff
@@ -418,4 +422,74 @@ func main() {
 			log.ErrorExit(err)
 		}
 	}
+
+	if *flgBMCLogCollectorConfigMap {
+		// create bmc-log-collector configmap of all servers that include boot servers
```
[the remark this answers is not in the document]
Comment (takara9) [reply machines-endpoints/pkg/machines-endpoints/main.go:427]: Deleted

File: machines-endpoints/pkg/machines-endpoints/main.go [hunk 6]
```diff
@@ -418,4 +422,74 @@ func main() {
...
+		addersses, err := updateServerAddressList(machines)
```
[the remark this answers is not in the document]
Comment (takara9) [reply machines-endpoints/pkg/machines-endpoints/main.go:428]: changed createMachinesList()

File: machines-endpoints/pkg/machines-endpoints/main.go (hunk 2 above)
Comment (takara9) [reply machines-endpoints/pkg/machines-endpoints/main.go:439]: Moved before main()

File: machines-endpoints/pkg/machines-endpoints/main.go (hunk 3 above)
Comment (takara9) [reply machines-endpoints/pkg/machines-endpoints/main.go:440]: Changed MachinesList

File: machines-endpoints/machines-endpoints.yaml [hunk 7]
```diff
@@ -102,6 +102,7 @@ spec:
               args:
```
[the remark this answers is not in the document]
Comment (takara9) [reply machines-endpoints/machines-endpoints.yaml:102]: THX

File: machines-endpoints/pkg/machines-endpoints/main.go [hunk 8]
```diff
@@ -346,6 +350,72 @@ func localHTTPClient() *http.Client {
...
+	if err != nil {
+		return err
+	}
+
+	configMap := corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: bmcLogCollectorConfigMapName,
+		},
+		Data: addresses,
+	}
```
Comment (masa213f) [inline machines-endpoints/pkg/machines-endpoints/main.go:404]: Variable name `addresses` is not appropriate.

```golang
	data := make(map[string]string)
	data["machineslist.json"] = machinesList

	ns, _, err := c.kubeConfig.Namespace()
	if err != nil {
		return err
	}

	configMap := corev1.ConfigMap{
		ObjectMeta: metav1.ObjectMeta{
			Name: bmcLogCollectorConfigMapName,
		},
		Data: data,
	}
```
or
```golang
	configMap := corev1.ConfigMap{
		ObjectMeta: metav1.ObjectMeta{
			Name: bmcLogCollectorConfigMapName,
		},
		Data: map[string]string{
			"machineslist.json": machinesList,
		},
	}
```

File: machines-endpoints/pkg/machines-endpoints/main.go [hunk 9]
```diff
@@ -418,4 +488,16 @@ func main() {
...
+	if *flgBMCLogCollectorConfigMap {
+		// create bmc-log-collector configmap of all servers
+		addersses, err := createMachinesList(machines)
+		if err != nil {
+			log.ErrorExit(err)
+		}
+		err = client.updateBmcLogCollectorConfigMap(ctx, addersses)
+		if err != nil {
+			log.ErrorExit(err)
+		}
```
Comment (masa213f) [inline machines-endpoints/pkg/machines-endpoints/main.go:501]: Variable name `addresses` is not appropriate.

```golang
		// create bmc-log-collector configmap of all servers
		machinesList, err := createMachinesList(machines)
		if err != nil {
			log.ErrorExit(err)
		}
		err = client.updateBmcLogCollectorConfigMap(ctx, machinesList)
		if err != nil {
			log.ErrorExit(err)
		}
```
or
```golang
		// create bmc-log-collector configmap of all servers
		str, err := createMachinesList(machines)
		if err != nil {
			log.ErrorExit(err)
		}
		err = client.updateBmcLogCollectorConfigMap(ctx, str)
		if err != nil {
			log.ErrorExit(err)
		}
```

**Review by masa213f — CHANGES_REQUESTED** [review]

File: machines-endpoints/pkg/machines-endpoints/main.go (hunk 8 above)
Comment (takara9) [reply machines-endpoints/pkg/machines-endpoints/main.go:404]: Changed

File: machines-endpoints/pkg/machines-endpoints/main.go (hunk 9 above)
Comment (takara9) [reply machines-endpoints/pkg/machines-endpoints/main.go:501]: Changed

File: machines-endpoints/pkg/machines-endpoints/main.go [hunk 10]
```diff
@@ -346,6 +351,69 @@ func localHTTPClient() *http.Client {
...
+			Name: bmcLogCollectorConfigMapName,
+		},
+		Data: data,
+	}
+
+	cms := c.k8s.CoreV1().ConfigMaps(ns)
+	currentConfigMap, err := cms.Get(ctx, bmcLogCollectorConfigMapName, metav1.GetOptions{})
+	if reflect.DeepEqual(nextConfigMap.Data, currentConfigMap.Data) {
+		return nil
+	}
```
Comment (masa213f) [inline machines-endpoints/pkg/machines-endpoints/main.go:404]: Please check errors before using `currentConfigMap`.

**Review by masa213f — CHANGES_REQUESTED** [review]

File: machines-endpoints/pkg/machines-endpoints/main.go (hunk 10 above)
Comment (takara9) [reply machines-endpoints/pkg/machines-endpoints/main.go:404]: Changed this
```
if reflect.DeepEqual(nextConfigMap.Data, currentConfigMap.Data) || err == nil {
```

Comment (takara9) [reply machines-endpoints/pkg/machines-endpoints/main.go]: changed

**Review by masa213f — APPROVED** [review]
LGTM

**Pull request merged**